File: widgetssdk/src/main/java/com/glia/widgets/dialog/domain/IsShowOverlayPermissionRequestDialogUseCase.java
package com.glia.widgets.dialog.domain;

import com.glia.widgets.dialog.PermissionDialogManager;
import com.glia.widgets.permissions.PermissionManager;

public class IsShowOverlayPermissionRequestDialogUseCase {
    private final PermissionDialogManager permissionDialogManager;
    private final PermissionManager permissionManager;

    public IsShowOverlayPermissionRequestDialogUseCase(
            PermissionManager permissionManager,
            PermissionDialogManager permissionDialogManager
    ) {
        this.permissionManager = permissionManager;
        this.permissionDialogManager = permissionDialogManager;
    }

    public boolean execute() {
        return hasNoOverlayPermissions() && hasNotShownOverlayPermissionRequest();
    }

    private boolean hasNoOverlayPermissions() {
        return !permissionManager.hasOverlayPermission();
    }

    private boolean hasNotShownOverlayPermissionRequest() {
        return !permissionDialogManager.hasOverlayPermissionDialogShown();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/internal/dialog/domain/IsShowOverlayPermissionRequestDialogUseCase.kt
package com.glia.widgets.internal.dialog.domain

import com.glia.widgets.internal.dialog.PermissionDialogManager
import com.glia.widgets.internal.permissions.PermissionManager
import com.glia.widgets.launcher.ConfigurationManager

internal interface IsShowOverlayPermissionRequestDialogUseCase {
    operator fun invoke(): Boolean
}

internal class IsShowOverlayPermissionRequestDialogUseCaseImpl(
    private val permissionManager: PermissionManager,
    private val permissionDialogManager: PermissionDialogManager,
    private val gliaSdkConfigurationManager: ConfigurationManager
) : IsShowOverlayPermissionRequestDialogUseCase {
    private val hasNoOverlayPermissions: Boolean get() = !permissionManager.hasOverlayPermission()
    private val hasNotShownOverlayPermissionRequest: Boolean get() = !permissionDialogManager.hasOverlayPermissionDialogShown()
    private val isUseOverlay: Boolean get() = gliaSdkConfigurationManager.enableBubbleOutsideApp

    override fun invoke(): Boolean = hasNoOverlayPermissions && hasNotShownOverlayPermissionRequest && isUseOverlay
}
